Source/tmModel/wnlib/mem/wnmem.man
NAME
     wnmem -- grouping memory allocator

SYNOPSIS
     #include "wnmem.h"

     void wn_gpmake(char parms[])

     void wn_gpfree(void)

     void wn_gppush(wn_memgp group);

     void wn_gppop(void)

     wn_memgp wn_curgp(void)

     wn_memgp wn_defaultgp(void)

     ptr wn_alloc(int size)

     ptr wn_zalloc(int size)

     void wn_realloc(ptr &p,int old_size,int new_size)

     void wn_zrealloc(ptr &p,int old_size,int new_size)

     void wn_free(ptr p)

     void wn_gperrfpush(void (*perror_func)(void))

     void wn_gperrfpop(void)

     typedef void *(*wn_system_memory_alloc_func_type)(unsigned int);
     typedef void  (*wn_system_memory_free_func_type)(ptr);

     void wn_set_system_memory_alloc_func(wn_system_memory_alloc_func_type)
     void wn_set_system_memory_free_func(wn_system_memory_free_func_type)

DESCRIPTION
     This package is a memory allocator designed to replace
     the malloc(3) package provided by UNIX.  It is designed
     to be faster and easier to program.

     All memory comes from a "memory group", which can be declared
     using the type "wn_memgp".  wnmem keeps a stack of memory
     groups.  The top of this stack is called the "current memory
     group" and can be accessed using "wn_curgp".  All memory is
     allocated from and freed to the current memory group.  The
     current memory group is NULL until a "wn_gppush" or "wn_gpmake"
     is done.
      
     "wn_gppush" pushes "group" onto the memory group stack.

     "wn_gppop" pops the memory group stack.

     "wn_gpmake" makes a memory group of type specified by "parms"
     and pushes it onto the memory group stack.  "parms" is a string 
     with the following syntax:

       parms --> free_method chunk_size

       free_method --> "no_free"|"general_free" 

       chunk_size --> <non-negative integer>|<empty>

     "free method" controls how "wn_free" works.
     "no_free" means that "wn_free" does nothing.  "no_free" groups
     run much faster and waste less space.  "general_free" groups
     allow piecemeal freeing with "wn_free".  Note that "wn_gpfree" works for
     both types of groups.  "chunk_size" is the size of the memory
     chunk that wnmem requests from the system.  The default is
     4000, which is a good choice unless you have many small groups.
     Giving 0 causes the smallest chunk possible to be requested.

     Some example wn_gpmake calls follow:

       wn_gpmake("no_free");
       wn_gpmake("general_free");
       wn_gpmake("no_free 100");
       wn_gpmake("no_free 0");

     The memory group created by "wn_gpmake" is a subgroup of
     the current group.  If a group is freed, all of its
     subgroups are freed.  If the current group is NULL,
     the memory group is a subgroup of no other group.

     "wn_gpfree" frees all of the memory in the current memory group,
     recursively frees its subgroups, invalidates the group,
     and then pops it from memory group stack.

     "wn_alloc" allocates "size" bytes from the current memory
     group and returns a pointer to it.

     "wn_zalloc" allocates "size" bytes from the current memory
     group and returns a pointer to it.  The memory is guaranteed
     to be null initialized.

     "wn_free" frees memory pointed to by "p" into the current memory
     group.  This group must be the same group that "p" was allocated
     from or trouble will result.  "wn_free" only works for "general_free"
     memory groups; it will do nothing for a "no_free" group.

     "wn_realloc" changes the size of a block of memory referenced by "p".
     "p" may be moved.

     "wn_zrealloc" does the same as "wn_realloc" except that all extra
     memory allocated will be null initialized.

     "wn_curgp" returns the current memory group.

     "wn_defaultgp" returns the "default" memory group, which
     is a no_free group used for various system purposes.

     "wn_set_system_memory_alloc_func" and
     "wn_set_system_memory_free_func" set new functions that wnmem is to
     call for its internal allocation and freeing.  The function pointers
     to be passed to these functions are to have the same argument and return
     types as malloc() and free().  If not called, wnmem uses malloc() and
     free().  Once these are called, all allocation for all groups goes
     through the new routines -- it recommended to call them before
     wnmem is initialized.

DIAGNOSTICS
     If memory is exhausted, the function provided to "wn_gperrfpush"
     is called.  By default, this function prints a message and
     then crashes.

     "wn_alloc", "wn_zalloc", and "wn_free" crash if the current
     memory group is NULL.
  
BUGS

SEE ALSO
     wnmemd

     doc/papers/memory_paper.txt

     cc/low/examples.c

AUTHOR
     Will Naylor




change_log.txt
/*******************************************************************************

5.0.1.20060413
- No changes, but now is release candidate.

20060325
- [RJL, all] Fixed more typos in the help docs.
- [CAF, linux] Updated wxWidgets patches and installation instructions.

20060323
- [RJL, all] Changed default symmetry angle for empty trees to 90 degrees.
- [RJL, all] Added discussion of color scheme to the View Settings window docs
and fixed a few typographical errors.
- [RJL, all] Changed RADIAN and DEGREES constants to be declared as constant
values rather than computed expressions, which fixed a problem in Intel Mac
builds (and only there) in which these constants were being given a value of 0.
- [RJL, all] Renamed Pi to PI, Two_Pi to TWO_PI for consistency and eliminated
const-ness of some return values for consistency.

5.0.0.20060309
- [RJL, all] Fixed a bug in the Path Combo inspector in which changing the
condition type wasn't always getting applied.
- [CAF, linux] - window fitting commands require WM hinting in some desktops;
make doc windows slightly less tall so bottom isn't hidden in some desktops.
- [CAF, all] Number pad delete key now deletes conditions in inspector (along
with backspace and fwd delete).
- [CAF, linux] Fixed missing double-click events in wxGTK list boxes
- [ABX, MSW] Printing dashing fixed.

20060308
- [CAF, Linux] Restored quit-on-close to match MSW behavior.

20060307
- [CAF, linux] invoke updating of user interface events in UpdateAllViews -
should patch the "unresponsive key after menu" issue in wxGTK
- [CAF, linux] like we did before in tmwxDocManager.cpp::OnFileClose (for closing
doc via menu/C-W), in tmwxDesignFrame::OnClose explicitly call base's method
instead of event.Skip(); [possibly a wxGTK, not Linux, issue]. Otherwise, when
a document window is killed via title bar button, the doc is not actually
closed, orphaning the inspector windows without creating a new empty doc.
- [RJL, all] Holding down modifier key when creating a new node by clicking
keeps the current selection.

20060304
- [ABX, MSW] Inspector and View Settings in on top when needed and on
background when unneeded. 
- [ABX, all] Added wrappers for range selection so it not influence existing
quality but is available for further development (just define
TM_WITH_RANGE_SELECTION macro in IDE/build). 
- [ABX, all] Assert during menu creation fixed.

20060228
- [RJL, all] Fixed a bug in tmConditionNodesCollinear in which deleting one of
the nodes could cause a crash.
- [RJL, all] Fixed an incorrect label in display of edges same strain condition.
- [RJL, all] In tmwxDesignCanvas::UpdatePaperSize, changed order of resizing
virtual image and resizing frame (frame now resized first) to avoid flicker
(that was only a problem on linux).

20060227
- [RJL, all] Rewrote Fit-to-[something] routines in tmwxDesignCanvas and
related supporting routines. Now supports user changes in monitor size and menu
items are enabled iff the command would alter the window position or size. Also
fixed several uninitialized members in tmwxDesignCanvas. Renamed several
tmwxDesignCanvas variables and routines.
- [RJL, all] Set paper size dialog is now initialized with current paper size.
- [RJL, all] Replaced a bunch of public member framework-related variables with
private members and public getters (to separate read/write access).
- [ABX, MSW+linux] Standard menu items are decorated with icons. (not on Mac.)
- [ABX, all] Drag & drop is supported.

20060226
- [RJL, all] Fixed a bug in the Node Fixed to Position dialog, in which checkbox
values were being ignored.
- [RJL, all] replacement of assorted ".size() == 0" with ".empty()" calls.
- [RJL, all] Eliminated superfluous arguments in tmwxTextCtrl::ValidateSizeT.
- [RJL, all] Fixed a bug in tmTree::AbsorbNode() in which mIsLeafNode and
mIsLeafPath flags were not getting updated properly in tmTree::AbsorbEdge.
- [ABX, all] Added check to Fit... commands so that they're not doubly executed
if the paper is already fit to the screen in one or the other direction.

20060224
- [ABX, all] Implemented View->Fit to Width and View->Fit to Height, added docs.
- [ABX, MSW] Added modified setup0.h (wxWidgets file) to turn off unnneeded
parts of wxW to reduce file size.
- [RJL, all] Replaced static arrays sCalcLocFns, sAddConditionFns, sSelectionFns
with singleton functions to delay initialization (resolving potential
intialization order issues). Also reimplemented debugging stringstreams with
initialize-on-usage.
- [RJL, all] Fixed an issue in node absorption that could cause a crash if a
crease pattern was present.

20060223
- [RJL, all] Fixed a bug in tmStubFinder::TriangulateTree in which encountering
of a non-splittable order-4+ tree (yes, it happens) caused a crash. Updated the
docs to note the possibility.
- [RJL, all] Fixed a bug in the display of a group selection (edges improperly
displayed). As part of that, implemented copy ctors and assignments in class

Source/tmModel/wnlib/mem/wnmem.c
/**********************************************************************

FUNCTIONS PROVIDED:

  wn_gpmake(parms)
  wn_gpfree()
  wn_gppush(group)
  wn_gppop()
  
  ptr wn_alloc(size)
  ptr wn_zalloc(size)
  wn_free(p)

  wn_realloc(&p,old_size,new_size)
  wn_zrealloc(&p,old_size,new_size)

  wn_memgp wn_curgp();
  wn_memgp wn_defaultgp();
  
  wn_gperrfpush(pfunc)
  wn_gperrfpop()  * not implemented *

  wn_make_gpstack(&stack)
  wn_get_current_gpstack(stack)
  wn_set_current_gpstack(stack)
  wn_swap_current_gpstack(stack)
  
DEBUGGING:

  wn_gp_fill();
  wn_gp_pad();
  wn_stack_fill();
  wn_gp_args_check();  * not implemented *
  wn_gp_trap_address(address);
  
  wn_allmem_verify();
  wn_group_verify(group);

  int wn_mem_used()
  int wn_group_mem_used(group)

  wn_print_mem_used()
  wn_print_all_groups_mem_used()
  wn_print_group_mem_used(group)

  bool wn_mem_in_group(p,group)

  wn_print_gp_stack()

  wn_gplabel(label)

**********************************************************************/
/****************************************************************************

COPYRIGHT NOTICE:

  The source code in this file is provided free of charge
  to the author's consulting clients.  It is in the
  public domain and therefore may be used by anybody for
  any purpose.

AUTHOR:

  Will Naylor

****************************************************************************/

#include <stdio.h>
#include <stdlib.h>

/*[RJL] Use standard library name under CW */
#ifdef __MWERKS__
  #include <ctype.h>
#else
/*[\RJL]*/
  #include <sys/types.h>
/*[RJL]*/
#endif
/*[\RJL]*/

#include <malloc.h>
#include <string.h>

#include "wnlib.h"
#include "wnasrt.h"
#include "wnmax.h"
#include "wnmemb.h"

#include "wnmem.h"



#define BEGIN_MAGIC      1122112211
#define END_MAGIC        1221122112

local bool initialized = FALSE;

local wn_gpstack current_gpstack = NULL;

local wn_memgp group_list = NULL;
local wn_memgp default_group = NULL;

bool wn_gp_fill_flag=FALSE,
     wn_gp_pad_flag=FALSE,
     wn_gp_trap_address_flag=FALSE;
ptr wn_gp_trap_address_address;

ptr wn_system_alloc(/*size*/);

local void default_error_func(int size);
typedef void (*voidfunc)(int size);
local voidfunc perror_func = (default_error_func);

void wn_initialize_group_for_general_free(wn_memgp group);
void wn_initialize_group_for_no_free(wn_memgp group);


local wn_system_memory_alloc_func_type lo_system_memory_alloc_func = &malloc;
local wn_system_memory_free_func_type  lo_system_memory_free_func  = &free;


void wn_set_system_memory_alloc_func(wn_system_memory_alloc_func_type func)
{
  lo_system_memory_alloc_func = func;
}


void wn_set_system_memory_free_func( wn_system_memory_free_func_type  func)
{
  lo_system_memory_free_func  = func;
}


ptr wn_system_alloc(int size)
{
  ptr ret;

  ret = (ptr)(*lo_system_memory_alloc_func)((unsigned int) size);

  if(ret == NULL)  /* out of memory */
  {
    (*perror_func)(size);  /* print out of memory message and crash,
                              or free up necessary memory */

    /* try again */
    ret = (ptr)(*lo_system_memory_alloc_func)((unsigned int) size);

    if(ret == NULL)
    {
      fprintf(stderr,"out of memory.\n");
      wn_assert_notreached();  /* no second chance */
    }
  }

  if(wn_gp_fill_flag)
  {
    wn_memset(ret,'a',size);
  }

  return(ret);
}


void wn_system_free(ptr mem,int size)
{
  if(wn_gp_fill_flag)
  {
    wn_memset(mem,'f',size); 
  }

  (*lo_system_memory_free_func)(mem);
}


local void print_group(wn_memgp group)
{
  char *label;

  if(group == NULL)
  {
    printf("%20s  %lx","",(unsigned long)group);
  }
  else
  {
    label = group->label;

    if(label == NULL)
    {
      label = "(NULL)";
    }

    printf("%20s  %lx",label,(unsigned long)group);
  }
}


void wn_print_gp_stack(void)
{
  int i;
  wn_memgp group;

  printf("************** memory group stack *************\n");

  printf("top -> ");    /* top of stack */
  print_group(current_gpstack->current_group);
  printf("\n");

  for(i=current_gpstack->stack_ptr;i>=0;i--)
  {
    group = (current_gpstack->group_stack)[i];

    printf("       "); 
    print_group(group);
    printf("\n");
  }

  printf("***********************************************\n");
}


void wn_gppush(wn_memgp group)
{
  ++(current_gpstack->stack_ptr);

  if (current_gpstack->stack_ptr >= GROUP_STACK_SIZE)
  {
    fprintf(stderr, "wnlib: ERROR: memory group stack overflow\n");
    wn_assert_notreached();
  }

  (current_gpstack->group_stack)[current_gpstack->stack_ptr] =
                                      current_gpstack->current_group;
  current_gpstack->current_group = group;
}


void wn_gppop(void)
{
  current_gpstack->current_group =
               (current_gpstack->group_stack)[current_gpstack->stack_ptr];

  --(current_gpstack->stack_ptr);

  /* it's an error to pop the last group (the default group) off the stack */
  if (current_gpstack->stack_ptr < 0)
  {
    fprintf(stderr, "wnlib: ERROR: memory group stack underflow\n");
    wn_assert_notreached();
  }
}


local void make_system_gpstack(wn_gpstack *pstack)    
{
  *pstack = (wn_gpstack)wn_system_alloc(sizeof(struct wn_gpstack_struct));

  (*pstack)->stack_ptr = -1;
  (*pstack)->group_stack = (wn_memgp *)
                        wn_system_alloc(GROUP_STACK_SIZE*sizeof(wn_memgp));
  (*pstack)->current_group = NULL;
}

/*[RJL] added prototype */
void wn_make_gpstack(wn_gpstack *pstack);
/*[\RJL]*/
void wn_make_gpstack(wn_gpstack *pstack)    
{
  *pstack = (wn_gpstack)wn_zalloc(sizeof(struct wn_gpstack_struct));

  (*pstack)->stack_ptr = -1;
  (*pstack)->group_stack = (wn_memgp *)
                               wn_zalloc(GROUP_STACK_SIZE*sizeof(wn_memgp));
  (*pstack)->current_group = NULL;
}


void wn_get_current_gpstack(wn_gpstack *pstack)
{
  *pstack = current_gpstack;
}


void wn_set_current_gpstack(wn_gpstack stack)
{
  current_gpstack = stack;
}


local void initialize_memory_allocator(void)
{
  make_system_gpstack(&current_gpstack);   

  initialized = TRUE;

  /*     make and push default memory group onto the bottom of the group
  ** stack.  It will be an error to pop it off. */
  wn_gppush(wn_defaultgp());
}


wn_memgp wn_curgp(void)
{
  if(!(initialized))
  {
    initialize_memory_allocator();
  }

  wn_assert(current_gpstack->stack_ptr >= 0);
  wn_assert(current_gpstack->stack_ptr < GROUP_STACK_SIZE);

  return(current_gpstack->current_group);
}


local void make_memory_group 
(
  wn_memgp *pgroup,
  int flags,
  int blocksize,
  wn_memgp parent_group
)
{
  *pgroup = (wn_memgp)wn_system_alloc(sizeof(struct wn_memgp_struct));

  (*pgroup)->begin_magic = BEGIN_MAGIC;
  (*pgroup)->end_magic = END_MAGIC;

  (*pgroup)->next = group_list;
  if(group_list != NULL)
  {
    group_list->plast = &((*pgroup)->next);
  }
  (*pgroup)->plast = &group_list;
  group_list = (*pgroup);

  if(parent_group != NULL)
  {
    (*pgroup)->sub_next = parent_group->children;
    if(parent_group->children != NULL)
    {
      parent_group->children->psub_last = &((*pgroup)->sub_next);
    }
    (*pgroup)->psub_last = &(parent_group->children);
    parent_group->children = (*pgroup);
  }
  else
  {
    (*pgroup)->sub_next = NULL;
    (*pgroup)->psub_last = NULL;
  }

  (*pgroup)->current_block = NULL;

# define DESIRED_SYSTEM_ALLOC_SIZE  8192
  (*pgroup)->block_size = DESIRED_SYSTEM_ALLOC_SIZE - 4 -
  /**/          sizeof(struct wn_mem_block_struct);
  (*pgroup)->block_ptr = NULL;
  (*pgroup)->block_mem_left = 0;
  (*pgroup)->mem_used = 0;

  (*pgroup)->pad_list = NULL;
  (*pgroup)->children = NULL;

  (*pgroup)->label = NULL;

  wn_assert(!(flags & WN_NO_FREE) != !(flags & WN_GENERAL_FREE));
  if (WN_NO_FREE & flags)
  {
    wn_initialize_group_for_no_free(*pgroup);
  }
  else if (WN_GENERAL_FREE & flags)
  {
    wn_initialize_group_for_general_free(*pgroup);
  }
  else
  {
    fprintf(stderr,"make_memory_group: illegal type of memory group\n");
    wn_assert_notreached();
  }

  if (blocksize)
  {
    (*pgroup)->block_size = blocksize;
  }
}


local void label_group(char *label,wn_memgp group)
{
  wn_assert(group->label == NULL);

  group->label = label;
}


void wn_gplabel(const char *label)
{
  char *label_copy;
  int save_mem_used;

  wn_assert(current_gpstack->stack_ptr >= 0);
   
  save_mem_used = current_gpstack->current_group->mem_used;

  label_copy = (char *)wn_alloc(strlen(label)+1);
  strcpy(label_copy,label);
  label_group(label_copy,current_gpstack->current_group);

  current_gpstack->current_group->mem_used = save_mem_used;
}


wn_memgp wn_defaultgp(void)
{
  if(!(initialized))
  {
    initialize_memory_allocator();
  }

  if(default_group == NULL)
  {
    make_memory_group(&default_group, WN_GENERAL_FREE, 4000, (wn_memgp)NULL);

    label_group("default_group",default_group);
  }

  return(default_group);
}


void wn_gpmake(const char parms[])
{
  char group_type[50];
  int group_size;
  int args;
  int flags;
  wn_memgp ret;

  if(!(initialized))
  {
    initialize_memory_allocator();
  }

  args = sscanf(parms, "%s %d", group_type, &group_size);
  wn_assert(args > 0  &&  args <= 2);

  if(strcmp(group_type,"no_free") == 0)
  {
    flags = WN_NO_FREE;
  }
  else if(strcmp(group_type,"general_free") == 0)
  {
    flags = WN_GENERAL_FREE;
  }
  else
  {
    fprintf(stderr,"wn_gpmake: illegal type of memory group\n");
    wn_assert_notreached();
  }

  if (2 > args)
  {
    group_size = 0;
  }

  make_memory_group(&ret, flags, group_size, current_gpstack->current_group);

  wn_gppush(ret);
}


void wn_gpmakef(int flags, int block_size)
{
  wn_memgp ret;

  if(!(initialized))
  {
    initialize_memory_allocator();
  }

  wn_assert(!(flags & WN_NO_FREE) != !(flags & WN_GENERAL_FREE));

  make_memory_group(&ret, flags, block_size, current_gpstack->current_group);

  wn_gppush(ret);
} /* wn_gpmakef */


local void free_group(wn_memgp group);

local void free_groups_children(wn_memgp group)
{
  wn_memgp child,next;

  next = group->children;

  for(;;)
  {
    if(next == NULL)
    {
      break;
    }

    child = next;
    next = child->sub_next;

    free_group(child);
  }
}


local void free_group_memory(wn_memgp group)
{
  wn_mem_block block,next;

  next = group->current_block;
  for(;;)
  {
    if(next == NULL)
    {
      break;
    }

    block = next;
    next = block->next;

    wn_system_free((ptr)block,block->size);
  }
}


local void remove_group_from_group_list(wn_memgp group)
{
  *(group->plast) = group->next;
  if(group->next != NULL)
  {
    group->next->plast = group->plast;
  }
}


local void remove_group_from_parent_subgroup_list(wn_memgp group)
{
  if(group->psub_last != NULL)
  {
    *(group->psub_last) = group->sub_next;
    if(group->sub_next != NULL)
    {
      group->sub_next->psub_last = group->psub_last;
    }
  }
}


local FILE *lo_trace_group_fp = NULL;
local long unsigned lo_last_heapsize = 0;
local int lo_traces_displayed = 0;

local void lo_trace_group_one_line(FILE *out, wn_memgp group)
{
  wn_mem_block block;
  long unsigned total_memory, leftover;
  long int free;
  double percent_wasted, percent_free, percent_leftover;
  char percent_free_buf[10];

  if (!lo_traces_displayed)
  {
    fprintf(out, "    %14s  %6s  %5s %9s %16s label\n", "total bytes",
    /**/      "%waste", "%free", "%leftover", "heapsize");
    fprintf(out, "    %14s  %6s  %5s %9s %16s ------------------\n",
    /**/  "--------------", "------", "-----", "---------",
    /**/    "----------------", "----------------");
  }

  if (!(lo_traces_displayed % 3))
  {
    fprintf(out, "\n");
  }

  total_memory = wn_total_memory_in_group(group);

  percent_wasted = 100.0 * (total_memory - wn_group_mem_used(group)) / 
  /**/              total_memory;

  free = wn_amount_of_free_memory_in_group(group);
  if (-1 == free)
  {
    strcpy(percent_free_buf, "no");
  }
  else
  {
    percent_free = 100.0 * free / total_memory;
    sprintf(percent_free_buf, "%6.2f", percent_free);
  }

  leftover = 0;
  for (block = group->current_block;  block;  block = block->next)
  {
    wn_assert(block->size > 0);
    leftover += block->leftover;
  }

  percent_leftover = 100.0 * leftover / total_memory;

  fprintf(out, "%3d %14lu %7.2f %6s %9.2f %16lu %s\n",
  /**/  lo_traces_displayed, total_memory, percent_wasted, percent_free_buf,
  /**/  percent_leftover, wn_heapsize(), group->label ? group->label : "");
  fflush(out);

  ++ lo_traces_displayed;
} /* lo_trace_group_one_line */


local void lo_trace_group_free_start(wn_memgp group)
{
  static bool first_time = TRUE;
  char *pfn;
  long unsigned new_heapsize;

  if (first_time)
  {
    first_time = FALSE;

    pfn = getenv("WNLIB_MEMGP_TRACE_FILE");
    if (pfn)
    {
      lo_trace_group_fp = fopen(pfn, "w");
      setbuf(lo_trace_group_fp, NULL);
    }
  }

  if (!lo_trace_group_fp)
  {
    return;
  }

  new_heapsize = wn_heapsize();
  if (new_heapsize < lo_last_heapsize)
  {
    fprintf(lo_trace_group_fp,
    /**/  "Drop at START of group free, by %13lu %14lu %s\n",
    /**/  lo_last_heapsize - new_heapsize, new_heapsize,
    /**/  group->label ? group->label : "");
  }
  lo_last_heapsize = new_heapsize;

  lo_trace_group_one_line(lo_trace_group_fp, group);
} /* lo_trace_group_free_start */


local void lo_trace_group_free_end()
{
  long unsigned new_heapsize;

  if (!lo_trace_group_fp)
  {
    return;
  }

  new_heapsize = wn_heapsize();
  if (new_heapsize < lo_last_heapsize)
  {
    fprintf(lo_trace_group_fp,
    /**/  "Drop at end   of group free, by %13lu %14lu\n",
    /**/  lo_last_heapsize - new_heapsize, new_heapsize);
  }
  lo_last_heapsize = new_heapsize;
} /* lo_trace_group_free_end */


local void free_group(wn_memgp group)
{
  free_groups_children(group);

  lo_trace_group_free_start(group);

  free_group_memory(group);

  remove_group_from_group_list(group);
  remove_group_from_parent_subgroup_list(group);

  wn_memset(group,'f',sizeof(struct wn_memgp_struct)); 

  wn_system_free((ptr)group,sizeof(struct wn_memgp_struct));

  lo_trace_group_free_end();
}


void wn_gpfree(void)
{
  free_group(current_gpstack->current_group);
  wn_gppop();
}


#define align(_size) \
{\
  (_size) = ((_size)+3)&(~3);\
}


ptr wn_zalloc(int size)
{
  ptr p;

  if(size == 0)
  {
    return(NULL);
  }
  if(!(initialized))
  {
    initialize_memory_allocator();
  }

  align(size);

  wn_assert(current_gpstack->stack_ptr >= 0);
  wn_assert(current_gpstack->stack_ptr < GROUP_STACK_SIZE);

  p = (*(current_gpstack->current_group->palloc_piece))(
                            size,current_gpstack->current_group);

  wn_memzero(p,size);

  return(p);
}


ptr wn_alloc(int size)
{
  if(size == 0)
  {
    return(NULL);
  }
  if(!(initialized))
  {
    initialize_memory_allocator();
  }

  align(size);

  wn_assert(current_gpstack->stack_ptr >= 0);
  wn_assert(current_gpstack->stack_ptr < GROUP_STACK_SIZE);

  return(
          (*(current_gpstack->current_group->palloc_piece))(
                            size,current_gpstack->current_group)
        );
}


void wn_free(ptr p)
{
  if(p == NULL)  /* zero size block from wn_alloc */
  {
    return;
  }

  wn_assert(current_gpstack->stack_ptr >= 0);
  wn_assert(current_gpstack->stack_ptr < GROUP_STACK_SIZE);

  (*(current_gpstack->current_group->pfree_piece))(
                                          p,current_gpstack->current_group);
}


void wn_realloc(ptr *pp,int old_size,int new_size)
{
  ptr new_p;

  new_p = wn_alloc(new_size);
  wn_memcpy(new_p,*pp,wn_min(old_size,new_size));
  wn_free(*pp);

  *pp = new_p;
}


void wn_zrealloc(ptr *pp,int old_size,int new_size)
{
  ptr new_p;

  new_p = wn_zalloc(new_size);
  wn_memcpy(new_p,*pp,wn_min(old_size,new_size));
  wn_free(*pp);

  *pp = new_p;
}


void wn_gperrfpush(void (*pfunc)(int size))
{
  perror_func = pfunc;
}


/*ARGSUSED*/ local void default_error_func(int size)
{
  fprintf(stderr,"out of memory\n");
  wn_assert_notreached();
}


void wn_gp_fill(void)
{
  wn_gp_fill_flag = TRUE;
}


void wn_gp_pad(void)
{
# if (WN_SIZEOF_PTR > WN_SIZEOF_INT)
    fprintf(stderr,
    /**/  "Warning: pads not implemented on 64-bit: wn_gp_pad() ignored\n");
# else
    wn_gp_pad_flag = TRUE;
# endif
}


local void stack_fill(int size)
{
  int filler[1000/sizeof(int)];

  if(size < 0)
  {
    return;
  }

  wn_memset(filler,'s',1000);

  stack_fill(size-1000);
}


void wn_stack_fill(void)
{
  stack_fill(100000);  /* fill 64k of stack */
}


void wn_gp_trap_address(ptr address)
{
  wn_gp_trap_address_flag = TRUE;
  wn_gp_trap_address_address = address;
}


local bool group_is_in_group_list(wn_memgp group)
{
  wn_memgp g;

  for(g=group_list;g != NULL;g=g->next)
  {
    if(group == g)  /* group is known */
    {
      return(TRUE);
    }
  }

  return(FALSE);
}


local void verify_group_stack_group(wn_memgp group)
{
  if(!(group_is_in_group_list(group)))
  {
    fprintf(stderr,"bad group in group stack.\n");
    wn_assert_notreached();
  }
}


local void generic_verify_group(wn_memgp group)
{
  wn_assert(group->begin_magic == (int) BEGIN_MAGIC);
  wn_assert(group->end_magic   == (int) END_MAGIC);
}


local void verify_group(wn_memgp group)
{
  generic_verify_group(group);

  (*(group->pverify_group))(group);
}


local void verify_group_stack(void)
{
  int i;
  wn_memgp group;

  if(current_gpstack == NULL)   /* no stack created yet */
  {
    return;
  }
  if(current_gpstack->stack_ptr == -1)  /* stack is empty */
  {
    return;
  }

  verify_group_stack_group(current_gpstack->current_group);

  for(i=current_gpstack->stack_ptr;i>0;i--) /* don't include 0, 0 always NULL */
  {
    group = (current_gpstack->group_stack)[i];

    verify_group_stack_group(group);
  }
}


local void verify_all_groups(void)
{
  wn_memgp group;

  for(group=group_list;group != NULL;group=group->next)
  {
    verify_group(group);
  }
}


/*[RJL] added prototype */
void wn_group_verify(wn_memgp group);
/*[\RJL] */
void wn_group_verify(wn_memgp group)
{
  if(!(group_is_in_group_list(group)))
  {
    fprintf(stderr,"bad group passed.\n");
    wn_assert_notreached();
  }

  verify_group(group);
}


void wn_allmem_verify(void)
{
  verify_group_stack();
  verify_all_groups();
}


local long int memory_used_in_group(wn_memgp group)
{
  return(group->mem_used);
}


long int wn_mem_used(void)
{
  long int total;
  wn_memgp group;

  total = 0;

  for(group=group_list;group != NULL;group=group->next)
  {
    total += memory_used_in_group(group);
  }

  return(total);
}


void wn_print_mem_used(void)
{
  char string[150];

  (void)sprintf(string,"total memory used = %d\n",wn_mem_used());
  fputs(string,stdout);
  fflush(stdout);
}


void wn_print_group_mem_used(wn_memgp group)
{
  char string[150];

  (void)sprintf(string,"total memory used for group = %d\n",
  /**/            wn_group_mem_used(group));
  fputs(string,stdout);
  fflush(stdout);
}


void wn_print_all_groups_mem_used(void)
{
  long int total_used;
  wn_memgp group;

  lo_traces_displayed = 0;

  total_used = 0;
  for(group = group_list; group != NULL; group = group->next)
  {
    lo_trace_group_one_line(stdout, group);

    total_used += wn_total_memory_in_group(group);
  }

  printf("--------------------------------------------------------------\n");
  printf("Total memory used = %22ld\n", total_used);
} /* wn_print_all_groups_mem_used */


long int wn_group_mem_used(wn_memgp group)
{
  long int total;

  total = memory_used_in_group(group);

  return(total);
}


/* total memory in all the blocks in the group, including wasted space.
** Does not include children of group. */

long unsigned wn_total_memory_in_group(wn_memgp group)
{
  long unsigned total = 0;
  wn_mem_block block;

  for (block = group->current_block;  block;  block = block->next)
  {
    wn_assert(block->size > 0);
    total += block->size;
  }

#if 0 /* ignore memory in children */
  for (child = group->children;  child;  child = child->sub_next)
  {
    total += wn_total_memory_in_group(child);
  }
#endif

  return total;
} /* wn_total_memory_in_group */


#if defined(__unix) && ! defined(lint)
#ifdef  __cplusplus
extern "C" {
#endif
extern char *sbrk(int);
extern int _end;
#ifdef  __cplusplus
}
#endif
#endif


long unsigned wn_heapsize(void)
{
    long unsigned mem;

#if defined(__unix) && ! defined(lint) && ! defined(__CYGWIN__)
    mem = (char *) sbrk(0) - (char *) &_end;
#else
    mem = 0;
#endif

    return mem;
}


local bool memory_is_in_block(ptr p,wn_mem_block block)
{
  long unsigned int start,fin,p_address;

  p_address = (long unsigned int)p;

  start = (long unsigned int)block;
  fin = start + block->size;

  return(
          (start <= p_address)
            &&
          (p_address < fin)
        );
}


bool wn_mem_in_group(ptr p,wn_memgp group)
{
  wn_mem_block block;

  for(block = group->current_block;
      block != NULL;
      block = block->next)
  {
    if(memory_is_in_block(p,block))
    {
      return(TRUE);
    }
  }

  return(FALSE);
}


local void
group_memory_fragmentation_report(wn_memgp group, int level)
{
  /* stub */
}


void
wn_memory_fragmentation_report()
{
  /* stub */
}
